File: String/791.Custom_Sort_String/C++/791.Custom_Sort_String.cpp
class Solution {
public:
    string customSortString(string order, string s) 
    {
        //============================//
        // Input type:                //
        //   - string (order string)  //
        //   - string (s string)      //
        // Return type:               //
        //   - string (result string) //
        //============================//

        /*Initialize*/
        ///// Length of strings (i.e. order, s) /////
        int len_order = order.size(), len_s = s.size();

        ///// Record dictionary, hashmap /////
        unordered_map<char, int> record_map;

        ///// Result string /////
        string res_str;


        /*Hashmap-based loop traversal*/
        ///// Step 1: Record indexed-char with dictionary, hashmap /////
        for (int s_idx = 0; (s_idx < len_s); (s_idx++)) //Whole
        {
            ((record_map[s[s_idx]])++); //Keep updating, recording

        } //Whole

        ///// Step 2: Looped-traversal for the recorded dictionary, hashmap /////
        for (int order_idx = 0; (order_idx < len_order); (order_idx++)) //Whole
        {
            while ((--(record_map[order[order_idx]])) >= 0) //Part
            {
                res_str += order[order_idx]; //Keep updating, recording

            } //Part

        } //Whole

        for (auto record_ele = record_map.begin(); (record_ele != record_map.end()); (record_ele++)) //Whole
        {
            while ((--(record_ele->second)) >= 0) //Part
            {
                res_str += (record_ele->first); //Keep updating, recording

            } //Part

        } //Whole

        return res_str;
    }
};

File: String/791.Custom_Sort_String/Python/791.Custom_Sort_String.py
class Solution(object):
    def customSortString(self, order, s):
        """
        :type order: str
        :type s: str
        :rtype: str
        """
        #================================#
        # Hashmap-based traversal method #
        #================================#

        ############
        #Initialize
        ##### Length of strings (i.e. order, s) #####
        len_order, len_s = len(order), len(s)

        ##### Record dictionary/hashmap #####
        record_dict = {}

        ##### Result string #####
        res_str = ""


        ##############################
        #Hashmap-based loop traversal
        ##### Step 1: Record indexed-char with dictionary/hashmap #####
        for s_idx in range(len_s):

            ##### Check if the current indexed-char existed or not #####
            if (s[s_idx] not in record_dict):
                record_dict[s[s_idx]] = 1 #Keep updating/recording
            else:
                record_dict[s[s_idx]] += 1 #Keep updating/recording

        ##### Step 2: Looped-traversal for the recorded dictionary/hashmap #####
        for order_idx in range(len_order):

            ##### Check if the current indexed-char existed or not #####
            if (order[order_idx] not in record_dict):
                continue

            while (record_dict[order[order_idx]] > 0):
                res_str += order[order_idx] #Keep updating/recording

                record_dict[order[order_idx]] -= 1 #Keep updating/reducing

        for record_key in record_dict:

            ##### Check if the current summary counter matched conditions or not #####
            if (record_dict[record_key] <= 0):
                continue

            while (record_dict[record_key] > 0):
                res_str += record_key #Keep updating/recording

                record_dict[record_key] -= 1 #Keep updating/reducing

        return res_str
